File: DataInjectorAndroid/DataInjectorLib/src/main/java/com/crescentflare/datainjector/transformer/StringifyTransformer.java
package com.crescentflare.datainjector.transformer;

import com.crescentflare.datainjector.conversion.InjectorConv;
import com.crescentflare.datainjector.injector.DataInjector;
import com.crescentflare.datainjector.utility.InjectorPath;
import com.crescentflare.datainjector.utility.InjectorResult;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.List;
import java.util.Map;

/**
 * Data injector transformation: transform data to a string
 * Convert data to a string, including nested structures
 */
public class StringifyTransformer extends BaseTransformer
{
    // --
    // Members
    // --

    private InjectorPath sourceDataPath;
    private boolean includeSpaces;
    private boolean includeNewlines;


    // --
    // Initialization
    // --

    public StringifyTransformer()
    {
    }


    // --
    // Manual transformation
    // --

    @NotNull
    public static InjectorResult stringify(@Nullable Object data)
    {
        return stringify(data, false, false);
    }

    @NotNull
    public static InjectorResult stringify(@Nullable Object data, boolean includeSpaces)
    {
        return stringify(data, includeSpaces, false);
    }

    @NotNull
    public static InjectorResult stringify(@Nullable Object data, boolean includeSpaces, boolean includeNewlines)
    {
        return InjectorResult.withModifiedObject(dataToString(data, includeSpaces, includeNewlines, 0));
    }


    // --
    // General transformation
    // --

    @Override
    @NotNull
    protected InjectorResult onApply(@Nullable Object sourceData)
    {
        Object useSourceData = DataInjector.get(sourceData, sourceDataPath != null ? sourceDataPath : new InjectorPath());
        return stringify(useSourceData, includeSpaces, includeNewlines);
    }


    // --
    // Set values
    // --

    public void setSourceDataPath(@Nullable InjectorPath sourceDataPath)
    {
        this.sourceDataPath = sourceDataPath;
    }

    public void setIncludeSpaces(boolean includeSpaces)
    {
        this.includeSpaces = includeSpaces;
    }

    public void setIncludeNewlines(boolean includeNewlines)
    {
        this.includeNewlines = includeNewlines;
    }


    // --
    // Internal conversion
    // --

    private static String dataToString(Object data, boolean includeSpaces, boolean includeNewlines, int level)
    {
        if (data instanceof Map)
        {
            return mapToString((Map)data, includeSpaces, includeNewlines, level);
        }
        else if (data instanceof List)
        {
            return listToString((List)data, includeSpaces, includeNewlines, level);
        }
        else if (data != null)
        {
            String result = InjectorConv.asString(data);
            if (result == null)
            {
                result = data.toString();
            }
            if (result != null)
            {
                if (data instanceof String)
                {
                    return "\"" + result + "\"";
                }
                return result;
            }
        }
        return "null";
    }

    private static String mapToString(Map<?, ?> map, boolean includeSpaces, boolean includeNewlines, int level)
    {
        StringBuilder result = new StringBuilder("{");
        boolean firstElement = true;
        if (includeNewlines)
        {
            result.append('\n');
        }
        for (Object key : map.keySet())
        {
            String stringKey = dataToString(key, includeSpaces, includeNewlines, 0);
            String stringValue = dataToString(map.get(key), includeSpaces, includeNewlines, level + 1);
            if (!firstElement)
            {
                result.append(',');
                if (includeNewlines)
                {
                    result.append('\n');
                }
                else if (includeSpaces)
                {
                    result.append(' ');
                }
            }
            if (includeNewlines)
            {
                for (int i = 0; i < level * 2 + 2; i++)
                {
                    result.append(' ');
                }
            }
            result.append(stringKey);
            result.append(':');
            if (includeSpaces)
            {
                result.append(' ');
            }
            result.append(stringValue);
            firstElement = false;
        }
        if (includeNewlines)
        {
            if (!firstElement)
            {
                result.append('\n');
            }
            for (int i = 0; i < level * 2; i++)
            {
                result.append(' ');
            }
        }
        result.append('}');
        return result.toString();
    }

    private static String listToString(List<?> list, boolean includeSpaces, boolean includeNewlines, int level)
    {
        StringBuilder result = new StringBuilder("[");
        boolean firstElement = true;
        if (includeNewlines)
        {
            result.append('\n');
        }
        for (Object item : list)
        {
            String stringItem = dataToString(item, includeSpaces, includeNewlines, level + 1);
            if (!firstElement)
            {
                result.append(',');
                if (includeNewlines)
                {
                    result.append('\n');
                }
                else if (includeSpaces)
                {
                    result.append(' ');
                }
            }
            if (includeNewlines)
            {
                for (int i = 0; i < level * 2 + 2; i++)
                {
                    result.append(' ');
                }
            }
            result.append(stringItem);
            firstElement = false;
        }
        if (includeNewlines)
        {
            if (!firstElement)
            {
                result.append('\n');
            }
            for (int i = 0; i < level * 2; i++)
            {
                result.append(' ');
            }
        }
        result.append(']');
        return result.toString();
    }
}

File: DataInjectorIOS/DataInjector/Classes/transformer/StringifyTransformer.swift
//
//  StringifyTransformer.swift
//  DataInjector Pod
//
//  Library transformer: transform data to a string
//  Convert data to a string, including nested structures
//

import Foundation

/// A transformer converting simple data or nested structures to a string
open class StringifyTransformer: BaseTransformer {
    
    // --
    // MARK: Members
    // --
    
    public var sourceDataPath: InjectorPath?
    public var includeSpaces = false
    public var includeNewlines = false

    
    // --
    // MARK: Initialization
    // --
    
    public override init() {
    }

    
    // --
    // MARK: Manual transformation
    // --
    
    public static func stringify(fromData: Any?, includeSpaces: Bool = false, includeNewlines: Bool = false) -> InjectorResult {
        return InjectorResult(withModifiedObject: dataToString(data: fromData, includeSpaces: includeSpaces, includeNewlines: includeNewlines, level: 0))
    }

    
    // --
    // MARK: General injection
    // --
    
    open override func appliedTransformation(sourceData: Any?) -> InjectorResult {
        let useSourceData = DataInjector.get(from: sourceData, path: sourceDataPath ?? InjectorPath(path: ""))
        return StringifyTransformer.stringify(fromData: useSourceData, includeSpaces: includeSpaces, includeNewlines: includeNewlines)
    }


    // --
    // MARK: Internal conversion
    // --

    private static func dataToString(data: Any?, includeSpaces: Bool, includeNewlines: Bool, level: Int) -> String {
        if let dict = data as? [AnyHashable: Any?] {
            return dictToString(dict: dict, includeSpaces: includeSpaces, includeNewlines: includeNewlines, level: level)
        } else if let array = data as? [Any?] {
            return arrayToString(array: array, includeSpaces: includeSpaces, includeNewlines: includeNewlines, level: level)
        } else if data != nil {
            var result = InjectorConv.asString(value: data)
            if result == nil {
                result = data.debugDescription
            }
            if let stringResult = result {
                if data is String {
                    return "\"\(stringResult)\""
                }
                return stringResult
            }
        }
        return "null"
    }
    
    private static func dictToString(dict: [AnyHashable: Any?], includeSpaces: Bool, includeNewlines: Bool, level: Int) -> String {
        var result = "{"
        var firstElement = true
        if includeNewlines {
            result += "\n"
        }
        for key in dict.keys {
            let stringKey = dataToString(data: key, includeSpaces: includeSpaces, includeNewlines: includeNewlines, level: 0)
            let stringValue = dataToString(data: dict[key] ?? nil, includeSpaces: includeSpaces, includeNewlines: includeNewlines, level: level + 1)
            if !firstElement {
                result += ","
                if includeNewlines {
                    result += "\n"
                } else if includeSpaces {
                    result += " "
                }
            }
            if includeNewlines {
                for _ in 0..<level * 2 + 2 {
                    result += " "
                }
            }
            result += stringKey
            result += ":"
            if includeSpaces {
                result += " "
            }
            result += stringValue
            firstElement = false
        }
        if includeNewlines {
            if !firstElement {
                result += "\n"
            }
            for _ in 0..<level * 2 {
                result += " "
            }
        }
        result += "}"
        return result
    }
    
    private static func arrayToString(array: [Any?], includeSpaces: Bool, includeNewlines: Bool, level: Int) -> String {
        var result = "["
        var firstElement = true
        if includeNewlines {
            result += "\n"
        }
        for item in array {
            let stringItem = dataToString(data: item, includeSpaces: includeSpaces, includeNewlines: includeNewlines, level: level + 1)
            if !firstElement {
                result += ","
                if includeNewlines {
                    result += "\n"
                } else if includeSpaces {
                    result += " "
                }
            }
            if includeNewlines {
                for _ in 0..<level * 2 + 2 {
                    result += " "
                }
            }
            result += stringItem
            firstElement = false
        }
        if includeNewlines {
            if !firstElement {
                result += "\n"
            }
            for _ in 0..<level * 2 {
                result += " "
            }
        }
        result += "]"
        return result
    }

}
